### PR TITLE
Fix map column can not be non-nullable for java

### PR DESCRIPTION
**Pull request by res-life**
## Description
Make map column non-nullable for java.

Changes:
- Add a new method to pass nullable; Deprecate the old one.
- Update the tests.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
Update existing tests
- [x] The documentation is up to date with these changes.
Bug fix, no need to update document.

### CODE REVIEW

File: java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java [hunk 1]
```diff
@@ -522,7 +522,10 @@ protected String[] getFlatColumnNames(String[] ret) {
    * Maps are List columns with a Struct named 'key_value' with a child named 'key' and a child
    * named 'value'. The caller of this method doesn't need to worry about this as this method will
    * take care of this without the knowledge of the caller.
+   * Note: this method always returns a nullabe column, can not return non-nullable column.
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java:525]:
```suggestion
   *
   * Note: this method always returns a nullabe column, can not return non-nullable column.
```

File: java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java [hunk 2]
```diff
@@ -537,6 +540,30 @@ public static ColumnWriterOptions mapColumn(String name, ColumnWriterOptions key
...
 
+  /**
+   * Add a Map Column to the schema.
+   * <p>
+   * Maps are List columns with a Struct named 'key_value' with a child named 'key' and a child
+   * named 'value'. The caller of this method doesn't need to worry about this as this method will
+   * take care of this without the knowledge of the caller.
+   * @param isNullable is the returned map nullable.
+   * Note: If this map column is a key of another map, should pass isNullable = false.
+   * e.g.: map1(map2(int, int), int) the map2 should be non-nullable.
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java:551]:
```suggestion
   * Add a Map Column to the schema.
   *
   * Maps are LIST columns with a STRUCT named 'key_value' with a child named 'key' and a child
   * named 'value'. The caller of this method doesn't need to worry about this as this method will
   * take care of this without the knowledge of the caller.
   *
   * Note: If this map column is a key of another map, should pass isNullable = false.
   * e.g.: map1(map2(int, int), int) the map2 should be non-nullable.
   *
   * @param isNullable is the returned map nullable.
```

File: java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java [hunk 3]
```diff
@@ -537,6 +540,30 @@ public static ColumnWriterOptions mapColumn(String name, ColumnWriterOptions key
...
+   */
+  public static ColumnWriterOptions mapColumn(String name, ColumnWriterOptions key,
+                                              ColumnWriterOptions value, Boolean isNullable) {
+    StructColumnWriterOptions struct = structBuilder("key_value").build();
+    if (key.isNullable) {
+      throw new IllegalArgumentException("key column can not be nullable");
+    }
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java:558]:
```suggestion
    if (key.isNullable) {
      throw new IllegalArgumentException("key column can not be nullable");
    }
    StructColumnWriterOptions struct = structBuilder("key_value").build();
```

File: java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java [hunk 4]
```diff
@@ -537,6 +540,30 @@ public static ColumnWriterOptions mapColumn(String name, ColumnWriterOptions key
...
+    struct.childColumnOptions = new ColumnWriterOptions[]{key, value};
+    ColumnWriterOptions opt = listBuilder(name, isNullable)
```
Comment (revans2) [inline java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java:560]: I am confused why we need a new API for this? Can we not just update the old API? It looks like it was a missed requirement and making something not support nulls should be fine for maps.

**Review by revans2 — COMMENTED** [review]
Technically this is not a breaking change for CUDF, but because the plugin code fails if it is using a deprecated API it will break the spark plugin build.  Could you please have a PR ready to go in the Spark plugin so that the time we are broken is very short.

File: java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java (hunk 4 above)
Comment (revans2) [reply java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java:560]: Sorry forgot to delete this one when I understood the issue better.

**Comment by res-life** [comment]
Corresponding PR: https://github.com/NVIDIA/spark-rapids/pull/9147

File: java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java (hunk 1 above)
Comment (res-life) [reply java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java:525]: Done

File: java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java (hunk 2 above)
Comment (res-life) [reply java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java]: Done

File: java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java (hunk 3 above)
Comment (res-life) [reply java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java]: Done

**Review by revans2 — APPROVED** [review]
I would like to see the changes the Spark plugin PR approved before this is merged in.

**Comment by res-life** [comment]
ok to test

**Comment by res-life** [comment]
build

**Comment by res-life** [comment]
ok to test

**Comment by res-life** [comment]
rerurn tests

File: java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java [hunk 5]
```diff
@@ -522,7 +522,11 @@ protected String[] getFlatColumnNames(String[] ret) {
    * Maps are List columns with a Struct named 'key_value' with a child named 'key' and a child
    * named 'value'. The caller of this method doesn't need to worry about this as this method will
    * take care of this without the knowledge of the caller.
+   *
+   * Note: this method always returns a nullabe column, can not return non-nullable column.
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/ColumnWriterOptions.java:526]:
```suggestion
   * Note: This method always returns a nullabe column, cannot return non-nullable column.
```

**Review by revans2 — APPROVED** [review]

**Comment by res-life** [comment]
@gpucibot merge

**Pull request merged**